Firestore/core/src/firebase/firestore/util/comparator_holder.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_COMPARATOR_HOLDER_H
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_COMPARATOR_HOLDER_H
+
+#include <type_traits>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * A holder of some comparator (e.g. std::less or util::Comparator) that
+ * implements the empty base optimization for the common case where the
+ * comparator occupies no storage.
+ */
+template <typename C, bool = std::is_empty<C>::value>
+class ComparatorHolder {
+ protected:
+  ComparatorHolder(const C& comparator) noexcept : comparator_(comparator) {
+  }
+
+  const C& comparator() const noexcept {
+    return comparator_;
+  }
+
+ private:
+  C comparator_;
+};
+
+// Implementation to use when C is empty.
+template <typename C>
+class ComparatorHolder<C, true> : private C {
+ protected:
+  ComparatorHolder(const C& comparator) noexcept : C(comparator) {
+  }
+
+  const C& comparator() const noexcept {
+    return this;
```

Comment:
Should it be `return *this`?

Hmm, maybe I'm missing something, but it looks like it should be a compilation error, in which case I'd double-check whether this specialization gets selected (I think when map is instantiated in tests, the comparator is stateless).

Reply:
It should be *this, but there's no actual usage of this method yet so the compiler doesn't type check it. Adding a type error into the other variant doesn't cause a failure either.

(Note that in the dependent PR that implements insertion, this does get used and the compiler complains as expected.)